// silkedit/lib/qt_event_emitter.js
'use strict';
const util = require('util');
const EventEmitter = require('events');

module.exports = function(bridge) {
  function QtEventEmitter() {}

  QtEventEmitter.prototype.addListener = function(event, listener) {
    bridge.connect.call(this, event);
    EventEmitter.prototype.on.call(this, event, listener);
    return this;
  }

  QtEventEmitter.prototype.on = QtEventEmitter.prototype.addListener;

  QtEventEmitter.prototype.once = function(event, listener) {
    throw new Error("once is not supported");
  }
  
  QtEventEmitter.prototype._emit = function(event, ...args) {
    return EventEmitter.prototype.emit.call(this, event, ...args);
  }
  
  QtEventEmitter.prototype.emit = function(event, ...args) {
    bridge.emit.call(this, event, ...args);
    return EventEmitter.prototype.listenerCount.call(this, event) > 0;
  }

  QtEventEmitter.prototype.removeAllListeners = function(event) {
    EventEmitter.prototype.removeAllListeners.call(this, event);
    bridge.disconnect.call(this, event);
    return this;
  }

  QtEventEmitter.prototype.removeListener = function(event, listener) {
    EventEmitter.prototype.removeListener.call(this, event, listener);
    if (EventEmitter.prototype.listenerCount.call(this, event) == 0) {
      bridge.disconnect.call(this, event);
    } else {
      console.log(`EventEmitter.prototype.listenerCount.call(this, event): ${EventEmitter.prototype.listenerCount.call(this, event)}`);
    }
    return this;
  }

  util.inherits(QtEventEmitter, EventEmitter);

  return QtEventEmitter;
};
