--- ./src/breakpad_googletest_includes.h.origin	2015-11-15 17:42:04.166772200 +0900
+++ ./src/breakpad_googletest_includes.h	2015-11-16 18:34:50.737397000 +0900
@@ -30,8 +30,8 @@
 #ifndef BREAKPAD_GOOGLETEST_INCLUDES_H__
 #define BREAKPAD_GOOGLETEST_INCLUDES_H__
 
-#include "testing/gtest/include/gtest/gtest.h"
-#include "testing/include/gmock/gmock.h"
+//#include "testing/gtest/include/gtest/gtest.h"
+//#include "testing/include/gmock/gmock.h"
 
 // If AddressSanitizer is used, NULL pointer dereferences generate SIGILL
 // (illegal instruction) instead of SIGSEGV (segmentation fault).  Also,
